### DISCUSSION TITLE
Bug detected: app.post return bool in float format if the type of my pydantic object is Union[str,int,float,bool]. A fix it by Union[bool,str,int,float]

### DISCUSSION THREAD

**Discussion by dd-the-dd-sigma** [fastapi/fastapi#12060] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional, Union
from fastapi import FastAPI
from fastapi.testclient import TestClient
from pydantic import BaseModel
from importlib.metadata import version

# Obtenir les versions des packages
pydantic_version = version("pydantic")
fastapi_version = version("fastapi")

# Imprimer les versions
print(f"Pydantic version: {pydantic_version}")
print(f"FastAPI version: {fastapi_version}")
app = FastAPI()

class Case1(BaseModel):
    raw_value: Optional[Union[str, int, float, bool]] = None

class Case2(BaseModel):
    raw_value: Optional[Union[bool, str, int, float]] = None

class Case3(BaseModel):
    raw_value: Optional[Union[bool, str, int]] = None

@app.post("/case1")
async def not_working_route() -> Case1:
    return Case1(raw_value=True)

@app.post("/case2")
async def not_working_route() -> Case2:
    return Case2(raw_value=True)

@app.post("/case3")
async def not_working_route() -> Case3:
    return Case3(raw_value=True)


# Créer un client de test
client = TestClient(app)

# Tester les routes
response_1 = client.post("/case1")
response_2 = client.post("/case2")
response_3 = client.post("/case3")


# Afficher les résultats
print()
print("Response 1:", response_1.json())
print("Response 2:", response_2.json())
print("Response 3:", response_3.json())


# Vérifier les types
assert isinstance(response_1.json().get("raw_value"), bool)
assert isinstance(response_2.json().get("raw_value"), bool)
assert isinstance(response_3.json().get("raw_value"), bool)
```

### Description

Result of print : 
Pydantic version: 2.6.1
FastAPI version: 0.110.0

Response 1: {'raw_value': 1.0}
Response 2: {'raw_value': 1.0}
Response 3: {'raw_value': True}

Weirder then those result, in my more complex code, case 2 works, but not case 1, that is why I tought that was a bug

### Operating System

Windows

### Operating System Details

Windows

### FastAPI Version

0.110.0

### Pydantic Version

2.6.1

### Python Version

3.9.19

**Comment by dd-the-dd-sigma** [comment]
I would expect my boolean to stay in bool type

**Comment by YuriiMotov** [comment, marked as answer]
During the validation of `Union`s, FastAPI goes through all possible types of that `Union` and uses first of them that passes the validation.
So, you should order your types inside `Union` from most strict (`bool` in your case) to less strict (`str`)
